### PR TITLE
Move admin_audit hooks to proper event listeners

### PR DESCRIPTION
**Pull request by tcitworld**
Requires #35677

## Summary

Moves all of the hooks used by `admin_audit` to event listeners, where appropriate typed events already exist.

- Bases itself on  #35677 to have app management events
- Adds a new `ConsoleEventV2` typed event to replace the legacy `ConsoleEvent` event in OCP (already depreciated). However, this doesn't seem used anywhere else, so we could replace it straight away.
- Adds preview details on the `BeforePreviewFetchedEvent` event

Part of #14552. Follow #37194 to track all of the hooks left.

## TODO

- [ ] Add tests for the new listeners

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by come-nc** [comment]
Is that different/better than https://github.com/nextcloud/server/pull/32019 ?

**Comment by tcitworld** [comment]
I missed this one, thanks. :see_no_evil: 

Well, there's more stuff in mine. I'll simply take the LDAP stuff from Carl's PR, and take over https://github.com/nextcloud/server/pull/32018 as well.

File: apps/admin_audit/lib/Listener/SharingEventListener.php [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+			$this->shareDeleted($event);
+		}
+	}
+
+	private function shareCreated(ShareCreatedEvent $event) {
+		$share = $event->getShare();
+
+		$params = [
+			'itemType' => $share->getNodeType(),
+			'path' => $share->getNode()->getPath(),
```
Comment (tcitworld) [inline apps/admin_audit/lib/Listener/SharingEventListener.php:49]: **question:** Unfortunately, this makes extra requests to get the Node. Should the share hold this information?

Comment (come-nc) [reply apps/admin_audit/lib/Listener/SharingEventListener.php:49]: Extra PHP calls? Or extra database queries?
Does not seem that overkill to me.

File: lib/public/Console/ConsoleEventV2.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?php
+/**
```
Comment (come-nc) [inline lib/public/Console/ConsoleEventV2.php:2]: strict types

File: lib/public/User/Events/BeforeUserIdUnassignedEvent.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?php
+
```
Comment (come-nc) [inline lib/public/User/Events/BeforeUserIdUnassignedEvent.php:2]: strict types

**Comment by come-nc** [comment]
Please strict type the hell out of this, return types on all new methods and strict_type declare on all new files, otherwise nice cleanup!

File: apps/admin_audit/lib/Listener/SharingEventListener.php [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			'itemSource' => $share->getNodeId(),
+			'shareWith' => $share->getSharedWith(),
+			'permissions' => $share->getPermissions(),
+			'id' => $share->getId()
+		];
+
+		match ($share->getShareType()) {
```
Comment (blizzz) [inline apps/admin_audit/lib/Listener/SharingEventListener.php:59]: The type could also be parameterized, saving some lines. It's not translated after all.

File: apps/admin_audit/lib/Listener/UserManagementEventListener.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+use OCP\User\Events\UserChangedEvent;
+use OCP\User\Events\UserCreatedEvent;
+use OCP\User\Events\UserDeletedEvent;
+use OCP\User\Events\UserIdAssignedEvent;
+use OCP\User\Events\UserIdUnassignedEvent;
+
+/**
+ * @template-implements UserManagementEventListener<UserCreatedEvent|UserDeletedEvent|UserChangedEvent|PasswordUpdatedEvent|UserIdAssignedEvent|UserIdUnassignedEvent>
+ */
+class UserManagementEventListener extends Action implements IEventListener {
```
Comment (blizzz) [inline apps/admin_audit/lib/Listener/UserManagementEventListener.php:41]: there are more issues with the docblock, check the static scanner remarks

File: apps/user_ldap/lib/Jobs/Sync.php [hunk 6]
```diff
@@ -44,47 +46,72 @@
...
+		$this->avatarManager = $avatarManager;
+		$this->dbc = $dbConnection;
+		$this->ncUserManager = $ncUserManager;
+		$this->logger = $logger;
+		$this->userManager = $userManager;
+		$this->notificationManager = $notificationManager;
+		$this->mapper = $userMapping;
+		$this->connectionFactory = $connectionFactory;
+
+		$this->accessFactory = new AccessFactory(
```
Comment (blizzz) [inline apps/user_ldap/lib/Jobs/Sync.php:100]: It should be safe to inject the AccessFactory directly.

File: lib/private/Console/Application.php [hunk 7]
```diff
@@ -30,43 +30,48 @@
...
 
 	public function __construct(IConfig $config,
-								EventDispatcherInterface $dispatcher,
+								EventDispatcherInterface $legacyDispatcher,
+								IEventDispatcher $dispatcher,
 								IRequest $request,
 								LoggerInterface $logger,
-								MemoryInfo $memoryInfo) {
-		$defaults = \OC::$server->getThemingDefaults();
+								MemoryInfo $memoryInfo, Defaults $defaults) {
```
Comment (blizzz) [inline lib/private/Console/Application.php:71]: codes style, one argument per line pls

**Review by blizzz — CHANGES_REQUESTED** [review]
some remarks, but nothing fundamental

**Comment by come-nc** [comment]
@tcitworld Can I take over this one?

**Comment by tcitworld** [comment]
Yes please, I don't even want to have a look at the conflicts! :see_no_evil:

**Comment by come-nc** [comment]
Supplanted by https://github.com/nextcloud/server/pull/47865

**Pull request closed without merging**